File: code/test_suites/cpp/FunctionBuildTest/tests/test_bitarray_tobytes.cpp
#include <gtest/gtest.h>
#include "../src/function_bitarray_tobytes.cpp"

class TestConvertToBytesFunction : public ::testing::Test {
protected:
    ConvertToBytesFunction convert_to_bytes_func;
};

TEST_F(TestConvertToBytesFunction, TestConvertStandardBitarray) {
    std::string result = convert_to_bytes_func.convert_to_bytes("110101");
    EXPECT_EQ(result, "b'\\xd4'");
}

TEST_F(TestConvertToBytesFunction, TestConvertEmptyBitarray) {
    std::string result = convert_to_bytes_func.convert_to_bytes("");
    EXPECT_EQ(result, "b''");
}

TEST_F(TestConvertToBytesFunction, TestConvertPaddedBitarray) {
    std::string result = convert_to_bytes_func.convert_to_bytes("1101");
    EXPECT_EQ(result, "b'\\xd0'");
}

TEST_F(TestConvertToBytesFunction, TestConvertBitarrayWithTrailingZeros) {
    std::string result = convert_to_bytes_func.convert_to_bytes("1101000");
    EXPECT_EQ(result, "b'\\xd0'");
}

TEST_F(TestConvertToBytesFunction, TestNonBitarrayInput) {
    EXPECT_THROW(convert_to_bytes_func.convert_to_bytes("10102"), std::invalid_argument);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: code/test_suites/python/python/test_bitarray_tobytes.py
import unittest
from function_bitarray_tobytes import ConvertToBytesFunction

class TestConvertToBytesFunction(unittest.TestCase):

    def setUp(self):
        self.convert_to_bytes_func = ConvertToBytesFunction()

    def test_convert_standard_bitarray(self):
        ba = '110101' 
        result = self.convert_to_bytes_func.convert_to_bytes(ba)
        self.assertEqual(result, "b'\\xd4'")

    def test_convert_empty_bitarray(self):
        ba = ''
        result = self.convert_to_bytes_func.convert_to_bytes(ba)
        self.assertEqual(result, "b''") 

    def test_convert_padded_bitarray(self):
        ba = '1101'  
        result = self.convert_to_bytes_func.convert_to_bytes(ba)
        self.assertEqual(result, "b'\\xd0'") 

    def test_convert_bitarray_with_trailing_zeros(self):
        ba = '1101000'  
        result = self.convert_to_bytes_func.convert_to_bytes(ba)
        self.assertEqual(result, "b'\\xd0'")

    def test_non_bitarray_input(self):
        with self.assertRaises(TypeError):
            self.convert_to_bytes_func.convert_to_bytes("10102")  

if __name__ == "__main__":
    unittest.main()
